docs(how-to): add how-to migrate to uv (#2156)
canonical/charmcraft: docs/howto/migrate-plugins/charm-to-uv.rst

.. _howto-migrate-to-uv:

Migrate from the Charm plugin to the uv plugin
==============================================

For charms that use `uv`_, Charmcraft has a :ref:`craft_parts_uv_plugin`. Migrating
from the Charm plugin provides some benefits, such as using uv during the build

> process not having to maintain a separate ``requirements.txt`` file. If the

Reviewer: > not having to maintain a separate ``requirements.txt`` file

:partying_face:

charm to be migrated does not currently use uv, refer to the
`uv documentation <https://docs.astral.sh/uv/guides/projects/>`_ for instructions
on how to use uv for a Python project.

Update the project file
-----------------------

The first step is to update the project file to include the correct parts definition.
Depending on the history of a specific charm, it may not have an explicitly-included
``parts`` section determining how to build the charm. In this case, a ``parts`` section
can be created as follows:

.. code-block:: yaml
    :caption: charmcraft.yaml

    parts:
      my-charm:  # This can be named anything you want
        plugin: uv
        source: .

Add optional dependency groups

> ------------------------------

Reviewer: I recommend consolidating or reflowing this section and the next. It reads askew to start with an optional step about dependency groups and then discuss a second, required step about dependency groups.

Is the order critical?

If the charm has `dependency groups`_ that should be included when creating the virtual
environment, the ``uv-groups`` key can be used to include those groups when creating
the virtual environment.

.. note::
    This is useful and encouraged, though not mandatory, for keeping track of

>     library dependencies, as covered in the next section.

Reviewer 2: Not sure about the necessity of this. If we do need to mention that the key is encouraged but not mandatory, I feel that it could be worked into the previous paragraph rather than being in a note.

Reviewer: If it's useful and encouraged, it shouldn't be a note!

Include charm library dependencies
----------------------------------

Unlike the Charm plugin, the uv plugin does not install the dependencies for
included charmlibs. If any of the charm libraries used have ``PYDEPS``, these will
need to be added to the charm's dependencies, potentially as their own

> `dependency group <dependency groups_>`_.

Reviewer: We linked to this quite recently – generally not necessary to do it again so soon.

Author: Kept this link and removed the other one because I rearranged the sections a bit.

> To find these dependencies, check each library file for its ``PYDEPS``. A command
> that can find these is::
> 
>     find lib -name "*.py" -exec awk '/PYDEPS = \[/,/\]/' {} +
> 
> If run from the base directory of a charm, this will show all the PYDEPS declarations
> from all loaded charm libs. These can then be included in ``pyproject.toml``.

Reviewer: ```suggestion
To find these dependencies, check each loaded library file for its ``PYDEPS`` by running
the following command at the root of the charm project:

.. code-block:: bash

    find lib -name "*.py" -exec awk '/PYDEPS = \[/,/\]/' {} +

Next, in ``pyproject.toml``, list them in a ``charmlibs`` dependency group.
```

.. code-block:: toml
    :caption: pyproject.toml

    [dependency-groups]
    # Dependencies brought from libraries the charm uses.
    charmlibs = [
        "cosl",
        "pydantic",
        "cryptography",
        "ops>=2.0.0",
    ]

> Including this dependency group is as easy as adding it to ``charmcraft.yaml``:

Reviewer 2: ```suggestion
Including this dependency group is as easy as adding it to the project file:
```

Author: I very intentionally used `charmcraft.yaml` here to distinguish between the two project file's we're talking about (`charmcraft.yaml` and `pyproject.toml`) in this section.

.. code-block:: yaml
    :caption: charmcraft.yaml

>     :emphasize-lines: 5-6

Reviewer: :partying_face:

.. _howto-migrate-to-uv:

Migrate from the Charm plugin to the uv plugin
==============================================

For charms that use `uv`_, Charmcraft has a :ref:`craft_parts_uv_plugin`. This guide
shows how to migrate from the default Charm plugin to the uv plugin.

> Migrating from the Charm plugin provides some benefits, not having to maintain a
> separate ``requirements.txt`` file and using the much faster ``uv`` for package
> management rather than ``pip``.

Reviewer: ```suggestion
Migrating from the Charm plugin provides some benefits, notably not having to maintain a
separate ``requirements.txt`` file. For package management, uv is much faster than pip.
```

Reviewing this again to make the paragraph self-consistent.

can be created as follows:

.. code-block:: yaml
    :caption: charmcraft.yaml

    parts:
      my-charm:  # This can be named anything you want
        plugin: uv
        source: .

Include charm library dependencies
----------------------------------

Unlike the Charm plugin, the uv plugin does not install the dependencies for
included charmlibs. If any of the charm libraries used have ``PYDEPS``, these will
need to be added to the charm's dependencies, potentially as their own
`dependency group <dependency groups_>`_.

To find these dependencies, check each loaded library file for its ``PYDEPS`` by running
the following command at the root of the charm project:

.. code-block:: bash
    find lib -name "*.py" -exec awk '/PYDEPS = \[/,/\]/' {} +

Next, in ``pyproject.toml``, list them in a ``charmlibs`` dependency group.

.. code-block:: toml
    :caption: pyproject.toml

    [dependency-groups]
    # Dependencies brought from libraries the charm uses.
    charmlibs = [
        "cosl",
        "pydantic",
        "cryptography",
        "ops>=2.0.0",
    ]

Add dependency groups
---------------------

> If the charm has dependency groups, such as one for charm libraries, that should be
> included when creating the virtual environment, the ``uv-groups`` key can be used to
> include those groups when creating the virtual environment.

Reviewer: ```suggestion
If the charm has dependency groups that should be included when creating the virtual
environment, such as one for charm libraries, the ``uv-groups`` key can be set to
include them.
```

Yay, the diff worked.

Including this dependency group in your charm is as easy as adding it to

> ``charmcraft.yaml``:

Reviewer: With the new paragraph, this feels redundant. Cut?

Author: Don't want to fully cut (I kept it because without it the jump to the example felt a bit abrupt), but I'll work on a way to enbriefen and attach it to the previous paragraph.

Author: OK, I've adjusted it, and once again our line length limit is producing some uhh... "interesting" formatting.

I also added a brief bit after the example about ``uv-extras``. I chose not to add an example for that because it's the same format as groups, but I did link again to the reference page down here since I think they're likely to want to look at the reference from here again.

Reviewer: LGTM.
